### ISSUE TITLE
Accessing Results (aBSREL)

### ISSUE THREAD

**Issue by joannabaker** [veg/hyphy#1083]
I am running aBSREL on a series of sequence alignments. I have no issues running my model using either the CLI (batch input) or datamonkey version, and both produce a JSON file as expected that loads without issue. 

I am happy about the model results for the full adaptive model, and I think I can extract all of the necessary information from what is presented in the JSON file. However, one thing that I am interested in the individual estimates of dN and dS for each branch of the phylogenetic tree that are estimated in the baseline model. As far as I can see, I can get at the branch-wise estimates of omega from both the baseline and aBSREL model by downloading the fitted tree and examining branch lengths (assuming each branch length reflects the branch omega estimate?). However, I cannot see a way to access individual estimates of dN or dS. 

My questions:
1) Is there a way to modify the JSON output (or generate additional output files) that comprise this information? 

2) Alternatively, can I run a separate model that would recreate the initial "baseline" model step for aBSREL? What would be the right way to run this to ensure it is identical to the one implemented in aBSREL? 

Apologies if this is a naive question, my understanding of the batch language is only limited.

**Comment by spond** [comment]
Dear @joannabaker,

`aBSREL` only estimates **ratios** (dN/dS). These are recorded, for the baseline model in the `branch attributes` JSON object under the key `Baseline MG94xREV omega ratio` (see below).

If you want to estimate dN and dS trees (i.e. expected non-synonymous substitutions per non-synonymous site), you might consider using https://github.com/veg/hyphy-analyses/tree/master/FitMG94

Best,
Sergei

```
"branch attributes":{
   "0":{
     "BABOON":{
       "Baseline MG94xREV":0.001816497257696574,
       "Baseline MG94xREV omega ratio":0,
       "Corrected P-value":1,
       "Full adaptive model":0.001804208605887959,
       "LRT":0,
       "Nucleotide GTR":0.001679760658077364,
       "Rate Distributions":        [
[0, 1] 
        ],
       "Rate classes":1,
       "Uncorrected P-value":1,
       "original name":"BABOON"
      },
     "CAT":{
       "Baseline MG94xREV":0.2774236018822691,
       "Baseline MG94xREV omega ratio":1.622357111673185,
       "Corrected P-value":0.00717862275903397,
       "Full adaptive model":0.1211885151207007,
       "LRT":13.13907100860342,
       "Nucleotide GTR":0.2660755122040191,
       "Rate Distributions":        [
[0.3134756783773388, 0.5667784498470505],
        [5.944793599938124, 0.4332215501529495] 
        ],
       "Rate classes":2,
       "Uncorrected P-value":0.0004785748506022647,
       "original name":"CAT"
      },
...
```

**Comment by joannabaker** [comment]
Thank you for your very swift reply, @spond . I have tried to use this batch file but apparently need to update HyPhy before I can. Hopefully this can get me to where I need.  I have two things I'd like to clarify, if you don't mind.

Thank you for pointing me to the appropriate place in the JSON object. I understand that nodes are named according to the order of traversal - but I find this tricky to interpret. In my experience, some programs interpret nodes in different order to others, and so I expect it is worth labelling nodes. Given this, is there a node label convention/format that is HyPhy-friendliest?

Secondly, am I right in thinking that the baseline model in aBSREL is somewhat akin to the "free-ratio" model implemented in codeml (using a GTR + gamma codon model, with a corrected "F3X4" estimator, plus other differences such as treatment of ambiguities etc.)? If I understand from earlier reading about such local models implemented within HyPhy, my understanding is that resultant branch lengths will be measured in substitutions per nucleotide site - is this true?

**Comment by spond** [comment]
Dear @joannabaker,

By default, HyPhy names internal nodes in the post-order traversal order. The JSON file actually stores the input tree with these names explicitly indicated, e.g. see below.

```
"input":{
   "file name":"/Users/sergei/Development/hyphy/tests/hbltests/data/CD2.nex",
   "number of sequences":10,
   "number of sites":187,
   "partition count":1,
   "trees":{
     "0":"((((PIG:0.147969,COW:0.21343)Node3:0.08509899999999999,HORSE:0.165787,CAT:0.264806)Node2:0.058611,((RHMONKEY:0.002015,BABOON:0.003108)Node9:0.022733,(HUMAN:0.004349,CHIMP:0.000799)Node12:0.011873)Node8:0.101856)Node1:0.340802,RAT:0.050958,MOUSE:0.09795)"
    }
  },
```

You can also name your internal nodes explicitly when you supply the tree to the analysis, like so

```
(((A,B)AB,C)ABC...
```

2. The baseline is effectively the free ratio model **without any site-to-site rate variation** (this is the model from [Yang's lysin paper](https://www.ncbi.nlm.nih.gov/pubmed/11018152)). Branch lengths are indeed measured in subs/nucleotide site.

Best,
Sergei

**Comment by joannabaker** [comment]
Dear @spond, 

Thanks again for your swift and very helpful response. I've updated HyPhy and can run this new script and obtain what I think is all the output I need using interactive command-line input. However, I can't seem to create a batch file for automated use - I wonder if you might be able to identify what I might be doing wrong here? I have attached the batch file I've tried to use (along with various modifications) and below I'll paste the error.  I'm sure it's a very simple mistake I have made, I'm still trying to wrap my head around the input process. 

Thank you so much for your help, 

Jo 


[test_input.bf.txt](https://github.com/veg/hyphy/files/4235727/test_input.bf.txt)

**Error:**
Could not find source dataset file "PROMPT_FOR_FILE" (resolved to '/mnt/g/CF3x4')

I get the same error even when I remove the "CF3x4" input from the attached file.
